File: dashboard/src/models/project/translation.ts
import type { IApiUser } from '../user/user'
import type { IApiEntry } from './entry'
import type { IApiProject } from './project'
import type { IApiVote } from './vote'

export interface IApiTranslation {
  id: number
  created_at: Date
  approved: boolean
  author_id: number
  author?: IApiUser
  content: string
  entry_id: number
  entry?: IApiEntry
  language_id: number
  project_id: number
  project?: IApiProject
  votes?: IApiVote[]
  rating: number
  user_vote: -1 | 0 | 1
}

export interface INewTranslation {
  project_id: number
  entry_id: number
  language_id: number
  content: string
}

File: backend/app/models/translation.py
from database.db import db
from app.models.base_model import BaseModel
from app.config import TRANSLATION_CONTENT_MAX_LENGTH

class Translation(db.Model, BaseModel):
    __tablename__ = 'translation'
    
    project_id = db.Column(db.Integer, db.ForeignKey('project.id'), nullable=False)
    entry_id = db.Column(db.Integer, db.ForeignKey('entry.id'), nullable=False)
    approved = db.Column(db.Boolean, nullable=False, default=False)
    author_id = db.Column(db.Integer, db.ForeignKey('user.id'), nullable=False)
    content = db.Column(db.String(TRANSLATION_CONTENT_MAX_LENGTH), nullable=False)
    language_id = db.Column(db.Integer, db.ForeignKey('language.id'), nullable=False)
    votes = db.relationship('Vote', backref='translation', lazy=True, cascade='all, delete-orphan')

    def to_dict(self, follow=[]):
        data = super().to_dict(follow)
        data['rating'] = sum(v.is_upvote for v in self.votes) - sum(not v.is_upvote for v in self.votes)
        data['user_vote'] = 0
        for vote in self.votes:
            if vote.user_id == self.author_id:
                data['user_vote'] = 1 if vote.is_upvote else -1
                break

        return data